### ISSUE TITLE
[058] Match pretrained logits and independently audit tiny autograd gradients

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2-C#59]
**Prerequisite:** #58 and #1. **Budget:** 40–60 minutes of integration; reference/model execution is separate. This is the pretrained forward milestone gate.

## Scope
Create `tools/reference_hf.py`, `tests/test_hf_reference.c`, `tests/test_pretrained_logits.c`, and fixture manifests. The AI may provide the explained Python reference. Use pinned Transformers/PyTorch on CPU, eager attention, eval mode, dropout disabled, use_cache=False, one thread, no mixed precision. Set scale_attn_weights=True, scale_attn_by_inverse_layer_idx=False, reorder_and_upcast_attn=False, activation_function='gelu_new', layer_norm_epsilon=1e-5. Real-model forward uses float32. C uses the exported weights, not Python inference at runtime.

For real GPT-2 use B1, token IDs `[15496,11,616,1438,318,257,3303,13]`, and separately its first-token-only prefix. Save all logits for each tested position, final normalized hidden states, source checksums, package versions and generator hash under `models/reference/`. IDs are fixed independently of the tokenizer lessons. Compare **every** element: logits atol2e-3,rtol1e-3; final hidden states atol5e-4,rtol1e-3. Print maximum errors and the first failure. These are prescribed gates, not previously executed C results; investigate failures instead of silently widening them.

## Independent tiny autograd oracle
Load lesson024's exact556-scalar deterministic fixture into a tiny HF model with the inverse layout mapping. Convert the model to float64 **before** assigning the double fixture values; do not round the tiny double fixture through a float32 checkpoint. Use its inputs `[[0,1,2],[2,1,0]]` and already-shifted targets `[[1,2,3],[1,0,3]]`.

Obtain logits without passing labels to the model, then compute:
```
loss = torch.nn.functional.cross_entropy(
    logits.reshape(6, 7), targets.reshape(6), reduction='mean'
)
```
This uses all six target positions. Do not pass these targets as `labels=` to GPT2LMHeadModel's built-in loss, because that path would apply its own causal shift and check a different objective. Call backward on this explicit loss and export canonical parameter gradients, transposing projection gradients back to C layout. Export tied wte exactly once with both graph contributions.

Compare C double gradients with atol1e-6,rtol1e-4 and forward values with atol1e-10,rtol1e-9. Include a reference-side assertion of exactly six contributing targets and a test that the intentionally double-shifted objective differs. This checks manual backward against an independent derivative implementation, not only finite differences.

## Commands and tests
```
.venv/bin/python tools/reference_hf.py --tiny --out tests/fixtures/hf-reference
make test-double TEST=hf_reference
.venv/bin/python tools/reference_hf.py --source models/gpt2-source --out models/reference
make test-pretrained TEST=logits
make test
```
Register HF-reference and pretrained gates separately from dependency-free core tests; committed tiny fixtures may be checked offline when explicitly requested. Missing requested assets fail clearly. Repeat T1/T8 and causal-prefix comparisons; parameters and RNG remain unchanged. Do not use generated prose or argmax alone as the numerical oracle.

**Mutation:** skip the square projection transpose in a temporary export; pretrained/tiny parity must fail. Separately apply an extra target shift in the reference loss: its objective-contract test must fail. Restore both.

**Exit question:** Why do full-logit/gradient agreement provide stronger evidence than matching a few generated words, and why must both implementations differentiate exactly the same scalar objective?

Reference: https://huggingface.co/docs/transformers/v4.51.3/model_doc/gpt2 . Record maximum errors and provenance in `notes/step-058.md`; follow #1.
